File: 0146-lru-cache/0146-lru-cache.py
from typing import *
class ListNode:
    def __init__(self, key: int, val: int, previous = None, next = None):
        self.key = key
        self.val = val
        self.previous = previous
        self.next = next

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache: Dict[int, Optional[ListNode]] = {}
        self.head: ListNode = ListNode(-1, -1)
        self.tail: ListNode = ListNode(-2, -2)
        self.head.next = self.tail
        self.tail.previous = self.head

    def move_to_end(self, node: Optional[ListNode]):
        lastNode: ListNode = self.tail.previous

        lastNode.next = node
        node.previous = lastNode

        node.next = self.tail
        self.tail.previous = node

    def rewire_pointers(self, node: Optional[ListNode]):
        before: ListNode = node.previous
        after: ListNode = node.next

        before.next = after
        after.previous = before
        

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        
        node: ListNode = self.cache[key]
        value: int = node.val

        self.rewire_pointers(node)
        self.move_to_end(node)

        return value
        

    def put(self, key: int, value: int) -> None:

        node: Optional[ListNode] = None
        if key in self.cache:
            node = self.cache[key]
            node.val = value
            self.rewire_pointers(node)
        else:
            node = ListNode(key, value)
        self.cache[key] = node
        self.move_to_end(node)

        if len(self.cache) > self.capacity:
            lruNode: ListNode = self.head.next
            self.rewire_pointers(lruNode)
            del self.cache[lruNode.key]

        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 0146-lru-cache/0146-lru-cache.java
class Node {

    private int key;
    private int value;
    private Node previous;
    private Node next;

    Node(int key, int value) {
        this.key = key;
        this.value = value;
    }

    public int getKey() {return this.key;}
    public int getValue() {return this.value;}
    public void setKey(int key) {this.key = key;}
    public void setValue(int value) {this.value = value;}
    public Node getPreviousNode() {return this.previous;}
    public Node getNextNode() {return this.next;}
    public void setPreviousNode(Node previous) {this.previous = previous;}
    public void setNextNode(Node next) {this.next = next;}

}

class LRUCache {

    private int capacity;
    private Node head;
    private Node tail;
    private Map<Integer, Node> cache;

    public LRUCache(int capacity) {

        this.capacity = capacity;
        this.cache = new HashMap<>();
        this.head = new Node(-1, -1);
        this.tail = new Node(-1, -1);

        this.head.setNextNode(this.tail);
        this.tail.setPreviousNode(this.head);
        
    }

    private void moveToEnd(Node node) {
        Node lastNode = this.tail.getPreviousNode();

        lastNode.setNextNode(node);
        node.setPreviousNode(lastNode);

        node.setNextNode(this.tail);
        this.tail.setPreviousNode(node);
    }

    private void rewirePointers(Node node) {
        Node leftNode = node.getPreviousNode();
        Node rightNode = node.getNextNode();

        leftNode.setNextNode(rightNode);
        rightNode.setPreviousNode(leftNode);

        node.setNextNode(null);
        node.setPreviousNode(null);
    }
    
    public int get(int key) {

        Node node = this.cache.get(key);
        if (node == null) {return -1;}

        int value = node.getValue();

        this.rewirePointers(node);
        this.moveToEnd(node);

        return value;
        
    }
    
    public void put(int key, int value) {

        Node node = this.cache.get(key);

        if (node == null) {
            node = new Node(key, value);
        } else {
            node.setValue(value);
            this.rewirePointers(node);
        }

        this.cache.put(key, node);
        this.moveToEnd(node);

        while (this.cache.size() > this.capacity) {
            Node lruNode = this.head.getNextNode();
            this.rewirePointers(lruNode);
            this.cache.remove(lruNode.getKey());
        }


        
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */